pkg/v11/resource/tiller/tiller.go

```diff
@@ -0,0 +1,168 @@
+package tiller
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/errors/guest"
+	"github.com/giantswarm/helmclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
+	"github.com/giantswarm/tenantcluster"
+
+	"github.com/giantswarm/cluster-operator/pkg/cluster"
+	"github.com/giantswarm/cluster-operator/pkg/v10/key"
+)
+
+const (
+	Name = "tillerv11"
+)
+
+// Config represents the configuration used to create a new tiller resource.
+type Config struct {
+	BaseClusterConfig        cluster.Config
+	Logger                   micrologger.Logger
+	Tenant                   tenantcluster.Interface
+	ToClusterGuestConfigFunc func(obj interface{}) (v1alpha1.ClusterGuestConfig, error)
+}
+
+// Resource implements the tiller resource.
+type Resource struct {
+	baseClusterConfig        cluster.Config
+	logger                   micrologger.Logger
+	tenant                   tenantcluster.Interface
+	toClusterGuestConfigFunc func(obj interface{}) (v1alpha1.ClusterGuestConfig, error)
+}
+
+// New creates a new configured tiller resource.
+func New(config Config) (*Resource, error) {
+	if reflect.DeepEqual(config.BaseClusterConfig, cluster.Config{}) {
+		return nil, microerror.Maskf(invalidConfigError, "%T.BaseClusterConfig must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Tenant == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Tenant must not be empty", config)
+	}
+	if config.ToClusterGuestConfigFunc == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ToClusterGuestConfigFunc must not be empty", config)
+	}
+
+	r := &Resource{
+		baseClusterConfig:        config.BaseClusterConfig,
+		logger:                   config.Logger,
+		tenant:                   config.Tenant,
+		toClusterGuestConfigFunc: config.ToClusterGuestConfigFunc,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.ensureTillerInstalled(ctx, clusterGuestConfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.ensureTillerInstalled(ctx, clusterGuestConfig)
```

Comment:
Do we reuse tiller for all resources?

Reply:
We only use tiller in the chartoperator resource to manage the operator in the tenant cluster.

Reply:
The question should be: do we reuse tiller for all clusters?

Reply:
No each tenant cluster has its own tiller.  Due to the port forwarding I don't think they can be shared.

Reply:
So how the tiller is deleted after cluster deletion?

Reply:
How is netexporter be deleted on cluster deletion? Everything goes down with the tenant cluster, right?

Reply:
Yes tiller goes away with the tenant cluster. We don't delete it the same way we don't delete the giantswarm namespace in the tenant.

Reply:
Ack. I forgot it's installed in the tenant cluster.